settings/src/components/webauthn/register-key.js

```diff
@@ -1,84 +1,109 @@
 	const [ step, setStep ] = useState( 'input' );
-	// have state for the key name etc?, and it gets passed down as props?
-
-	const startRegistration = useCallback( () => {
-		// call preprgister endpoint to get options and nonce
-		// 	 why is this necessary? why not just ship it with the initial page build?
-		// handle failure
-		//
-		// call navigator.credentials.create with options from previous step
-		// handle failure
-		//
-
-		setStep( 'waiting' );
-	}, [] );
-
-	/**
-	 * Save the user's new security key in the database.
-	 *
-	 */
-	const saveKeyToDatabase = useCallback( () => {
-		// TODO make an API request to save the new key,
-		// then refreshRecord( userRecord );
-
-		// TODO the following is just a placeholder to demonstrate the flow until the above is implemented
-		const newKeys = userRecord.record[ '2fa_webauthn_keys' ].push( {
-			id: Math.random(),
-			name: 'New Key',
-		} );
-
-		userRecord.edit( { '2fa_webauthn_keys': newKeys } );
-
-		setStep( 'success' );
-	}, [ userRecord.record[ '2fa_webauthn_keys' ] ] );
-	// todo probably wont need userrecord dependency once the above is implemented
+	const [ registerCeremonyActive, setRegisterCeremonyActive ] = useState( false );
+
+	const onRegister = useCallback(
+		async ( event ) => {
+			try {
+				event.preventDefault();
+				setRegisterCeremonyActive( true );
+
+				const nonce = record[ '2fa_webauthn_register_nonce' ];
+				const preRegisterResponse = await wp.ajax.post( 'webauthn_preregister', {
+					user_id: record.id,
+					_ajax_nonce: nonce,
+				} );
+				const publicKey = preparePublicKeyCreationOptions( preRegisterResponse.options );
+
+				setStep( 'waiting' );
+
+				const credential = await navigator.credentials.create( { publicKey } );
+
+				await wp.ajax.post( 'webauthn_register', {
+					user_id: record.id,
+					_ajax_nonce: nonce,
+					name: keyName,
+					credential: JSON.stringify( preparePublicKeyCredential( credential ) ),
+				} );
+
+				await refreshRecord( userRecord );
+				setStep( 'success' );
+			} catch ( exception ) {
+				setError( exception?.message || exception?.responseJSON?.data || exception );
+				setStep( 'input' );
+			} finally {
+				setRegisterCeremonyActive( false );
+			}
+		},
+		[ keyName ]
+	);
 
 	if ( 'waiting' === step ) {
-		return <WaitingForSecurityKey onCeremonySuccess={ saveKeyToDatabase } />;
+		return <WaitingForSecurityKey />;
 	}
 
 	if ( 'success' === step ) {
-		return <Success newKeyName={ 'Test key' } afterTimeout={ onRegisterSuccess } />;
+		return <Success newKeyName={ keyName } afterTimeout={ onSuccess } />;
 	}
 
 	return (
-		<form>
-			<TextControl label="Give the security key a name" />
-			{ /* TODO add basic clientside validation
-			what? lendth? chars?
-			TextControl doesn't have anything special, just pass `maxlength`, `pattern`, etc
-			*/ }
+		<form onSubmit={ onRegister }>
+			<TextControl
+				label="Give the security key a name"
+				onChange={ ( name ) => setKeyName( name ) }
```

Comment:
Not sure if an empty keyName is allowed, for now it would be `NewKey` if nothing is input.
Probably some easy check on front-end side could be done in https://github.com/WordPress/wporg-two-factor/issues/194.

Reply:
Yeah, the API endpoint sets that default string here:

https://github.com/sjinks/wp-two-factor-provider-webauthn/blob/32c8c10f3f67a40a87e3a5f0cd643db568343bf7/inc/class-ajax.php#L121-L138

But I think it makes sense to to add a `required` attribute on our side. Done in ff14ce1 👍🏻